bolt-design-system/bolt
# Bolt Core Renderer + Rendering Animation Performance Comparison

## Jira
N/A

## Summary
Adds 3 new demos to Bolt to highlight major rendering performance differences + a potential new technique to help address issues with scale -- particularly necessary with upcoming animation work. 

## Details
- This PR adds a performance rendering demo of a [Sierpinski triangle](https://en.wikipedia.org/wiki/Sierpinski_triangle) implemented in both existing Bolt base components in comparison to potential performance gains by using Google's Lit Element + Lazy Lit Element POC.

  - [Lit-HTML](https://feature-animation-rendering-performance.boltdesignsystem.com/pattern-lab/patterns/01-styleguide-100-rendering-performance-bolt-lit-test-00-bolt-lit-test/01-styleguide-100-rendering-performance-bolt-lit-test-00-bolt-lit-test.html)
  - [Preact](https://feature-animation-rendering-performance.boltdesignsystem.com/pattern-lab/patterns/01-styleguide-100-rendering-performance-bolt-preact-test-02-bolt-preact-test/01-styleguide-100-rendering-performance-bolt-preact-test-02-bolt-preact-test.html)
  - [Lazy Lit Element (New)](https://feature-animation-rendering-performance.boltdesignsystem.com/pattern-lab/patterns/01-styleguide-100-rendering-performance-lazy-lit-test-lazy-lit-test/01-styleguide-100-rendering-performance-lazy-lit-test-lazy-lit-test.html)


This PR also updates a few things in the build tools related to getting these demos up and running:

1. Adds the ability to compile Typescript `.ts`, and `.tsx` files in Bolt's local Webpack Build
2. Updates the `@bolt/core` click event handler util to allow toggling JavaScript methods that are on the window (vs only allowing methods that are on an element itself)
3. Updates the `@bolt/core` base component to no longer immediately trigger a component without any props to render (via the `updated()` method call) -- fixes a minor issue discovered with the new Context API additions in #961 that can cause components using the new `withContext` functionality to have undefined values initially until the component finishes rendering. 

## How to test
- Confirm the build compiles as expected
- Confirm existing click event handlers (ex. buttons hooked up to start playing the video player on click) continue to work as expected
- Confirm the new rendering performance demos are working in modern browsers (Chrome, Safari and Firefox -- IE 11 support isn't yet possible due to the way slots are used in the demo components)


CC @mikemai2awesome @danielamorse @adam2661 -- particularly regarding the 3rd item above since I'm 99% sure Adam and I ran into this snag last week with the new List component work.
